<div align="left">
<img src="./docs/logo.png" width="300" height="auto">
</div>

<br/>
<b>A visualization demo can be seen here:</b> https://cbg-ethz.github.io/oncotreeVIS

# oncotreeVIS â€“ An interactive graphical user interface  for visualising mutation tree cohorts


### Abstract
----------

<p align="justify">Recent developments in single-cell omics technologies have revealed, with increasing precision, the intra- and inter-tumor heterogeneity, including the evolutionary history of tumors and their cell phylogeny, represented by mutation trees. Several mutation tree inference tools can reconstruct the evolutionary relationships among different tumor subclonal populations, which share common genomic events. The inspection of these mutation trees at the cohort level helps understanding the genomic heterogeneity of tumors, an important task in computational oncology and an essential factor for tumor board clinical decisions.</p>

<p align="justify">We propose <b>oncotreeVIS</b>, a command line tool and interactive graphical user interface for visualising mutation tree cohorts and tree posterior distributions obtained from tree inference tools. Our tool can be run locally for custom user input given in a predefined format, for mutation trees that encode single or joint genomic events such as point mutations, copy number changes or changes in gene expression levels and reflect different modes of tumor evolution governed by selection and by the spatial architecture of the tumor. By default we provide visualisations of several publicly available mutation tree cohorts. OncotreeVIS facilitates the visual inspection of mutation tree clusters and k-nearest neighbours around every mutation tree, together with their corresponding metadata, if available.</p>

<div align="center">
<img src="./docs/poster.png" width="90%" height="auto">
</div>
